### ISSUE TITLE
Add a bisection command that uses nightly builds

### ISSUE THREAD

**Issue by jdm** [servo/servo#19505]
http://servo-builds.s3.amazonaws.com/ hosts a long list of historical nightly builds. We should write a command that allows several modes of operation:
* `./mach bisect tests/wpt/web-platform-tests/foo.html` - runs WPT tests using downloaded nightly builds, suggests good/bad bisection based on test result
* `./mach bisect path/to/file.html` - runs downloaded nightlies and loads file:// url for easy testing
* `./mach bisect http://foo.com` - runs downloaded nightlies and loads url for easy testing
* `./mach bisect -b` - runs downloaded nightlies with browser.html for easy testing

We should cache the downloaded nightlies to make it easier to do future bisections, too.

**Comment by jdm** [comment]
`git rev-list -n 1 --before="2017-01-01" master` is a useful command to figure out a commit associated with a particular nightly.

**Comment by jdm** [comment]
Curiously, http://servo-builds.s3.amazonaws.com/ only shows mac builds up until 2016-10-25. There are plenty of linux and android builds for recent days, however.

**Comment by jdm** [comment]
Ah, `<MaxKeys>1000</MaxKeys>` is hiding the entries that I care about.

**Comment by jdm** [comment]
I use http://servo-builds.s3.amazonaws.com/?list-type=2&prefix=nightly/mac/2017 when investigating regression ranges for macOS builds from 2017.

**Comment by o0Ignition0o** [comment]
This one looks very interesting, could I please pick it? I think I can start working on it next week :)

**Comment by jdm** [comment]
That would be great!

**Comment by jdm** [comment]
@o0Ignition0o Have you made any progress with this? If not, I would like to turn this into a project for some students this term.

**Comment by o0Ignition0o** [comment]
Hi @jdm :) I haven't made too much progress with this yet, so you can assign it to someone else if you want to. I'll pick up an other task

**Comment by jdm** [comment]
That's fine; you're welcome to keep working on it!

**Comment by atouchet** [comment]
@jdm are Windows builds available with this? I see every other platform on there aside from Windows. Having historical nightly builds would be useful for testing purposes.

**Comment by jdm** [comment]
Note my comment earlier with the prefix argument. If you specify the right prefix, they should appear.

**Comment by atouchet** [comment]
http://servo-builds.s3.amazonaws.com/?list-type=2&prefix=nightly/windows-msvc/ appears to work, thanks.

**Comment by o0Ignition0o** [comment]
Hey there :) I've been able to write a rough try to download a (hardcoded for now) nightly version and extract it.

I put the bisect command as a "mach wrapper" for now so I can run any other mach command, eg : 
` ./mach bisect run https://www.mozilla.org ` 
` ./mach bisect help ` 
` ./mach test --all ` 

I'm now looking for a way to set the downloaded servo bin as the target I'm supposed to use.

I have a feeling this should be an environment variable, but I haven't found it quite yet ^^'
Next steps for me would be parsing s3's XML files (with beautifulsoup or something similar) and binary searching my way through the ./mach commands exit codes, but I need to be able to set the servo target first.
If you have an idea please let me know :)

**Comment by jdm** [comment]
https://github.com/servo/servo/blob/master/python/servo/post_build_commands.py#L99 is the implementation for `./mach run`, and it relies on [get_binary_path](https://github.com/servo/servo/blob/master/python/servo/command_base.py#L353). There's no built-in way to override the binary that is selected right now; adding an additional argument to the run command that passes a specific binary to use would probably make sense.

**Comment by o0Ignition0o** [comment]
Hi, thanks a lot for your response. 

So here's what I'm going to do : 

1 - Add a --bin flag to ./mach run to allow a user to specify the path to the executable to run

2 - Add a --nightly | -n flag to allow a user to specify a YYYY-MM-DD date. This flag will :
	Figure out which OS the user is running
	Check the nightlies folder to find out if the nightly version has already been downloaded and extracted
	if not : 
		Try to fetch it from the S3 server
		Raise an exception and print a message if there is no nightly for the specified date and platform
	run the specified command with the nightly version
    The default nightly folder will be PROJECT_ROOT/target/nightlies/ and should be customizable through an environment variable

3 - go back to the bisect command, which should be easier to implement once these steps are performed.

Each step will be a separate Pull Request, which I'll link to this issue. 
Should I file separate issues for each step ?
Are there things I should reconsider before working on the implementation ?

**Comment by jdm** [comment]
That sounds great!

**Comment by o0Ignition0o** [comment]
Just pushed a rough try to handle the --nightly flag.
I really don't like the way I let the -n flag silently fail if an empty string is provided, I'll amend my commit asap ^^'

**Comment by o0Ignition0o** [comment]
Ok so I chose to split my merge requests depending on the platforms, because I don't know how yo extract a .msi and a .dmg archive with python ^^'

If you have any hint pease let me know, I'll keep on digging :)

**Comment by o0Ignition0o** [comment]
Ok so it seems like passing the -qi flags to msiexec.exe (on windows) should allow me to silently install a servo build, but I'm not sure at all I wouldn't mess any registry keys / already installed versions of servo.
Could anyone please point me to someone knowledgeable on windows installers so I can get more insights ?

PS : In the mozilla-unified repository, we seem to use [executables](https://dxr.mozilla.org/mozilla-central/source/testing/mozbase/mozinstall/tests/test.py#50) so it seems easier to handle, but I'm not quite sure how I should handle .msi files :/

**Comment by o0Ignition0o** [comment]
Thanks a lot, I ll give it a try :)

**Comment by o0Ignition0o** [comment]
Just ran into github issues claiming running servo in a VM is not that easy, so I think the fact I can't run servo.exe might be related to my virtualbox setup.
I'll just remove the print occurrences and submit it for review.
I don't really know whether extracting only the executable is relevant or not, maybe as a followup bug if no other file is needed ?

**Comment by jdm** [comment]
If you add your ssh key to https://github.com/servo/saltfs/tree/master/admin/files/ssh and your username to https://github.com/servo/saltfs/blob/master/admin/map.jinja, you could get ssh access to our mac build machines if you wanted :)

**Comment by o0Ignition0o** [comment]
Oh this is great !
I'll file to get access then, thanks a lot !

**Comment by o0Ignition0o** [comment]
It looks like windows builds now produce an executable instead of a .msi archive, sothe -n flag doesnt work on windows anymore :( .

I'm trying to find a way to extract the executable without installing the application, but couldn't manage so far.

Also https://firefox-source-docs.mozilla.org/mozbase/mozinstall.html doesn't allow me to just extract a portable version, maybe I could submit a pr to mozinstall to allow people to just extract contents ?

**Comment by jdm** [comment]
Interesting point; that's an unfortunate consequence. If you want to change mozinstall, you probably need to submit a patch to mozilla-central, since I suspect that https://searchfox.org/mozilla-central/source/testing/mozbase/mozinstall is the authoritative source.

**Comment by o0Ignition0o** [comment]
Yeah, plus that would maybe allow us to use exclusively the mozinstall API on Windows Linux and macos, which would be an improvement. I'll try to submit a patch :)

**Comment by jschwe** [comment]
I believe we do already have a `--nightly` argument (as part of the `binary_selection` decorator, so implementing this might be easier today. Nightlies are now also mainly hosted on github releases: https://github.com/servo/servo-nightly-builds/releases
